### PR TITLE
Load Widget Extensions and support extensions such as PyAladin

### LINKED ISSUE #9292: IPyaladin widget does not work in VSCode
### Discussed in https://github.com/microsoft/vscode-jupyter/discussions/9290

<div type='discussions-op-text'>

<sup>Originally posted by **JosephKarpinski** March  8, 2022</sup>
Hi, 

I’m lost on this.
This jupyter notebook works fine in Jupyter but not in Visual Studio.
Nothing is displayed. Attaching error message.

D1985D1C-B788-4F0D-862B-4CC70347281D
55C0516D-26D4-4594-B4D9-DE69CB479B6D
21E98D11-AF81-4277-92AA-AC6078FF55ED
3ABC4FB5-B3D4-41F0-A18D-222B20161A3D
6EA74317-BC6A-45DC-A4F9-BC831910A28A
</div>

### PR DESCRIPTION
**Pull request by DonJayamanne**
Fixes #9292

### CODE REVIEW

File: src/kernels/helpers.ts [hunk 1]
```diff
@@ -1302,10 +1302,10 @@ export function getLanguageInNotebookMetadata(metadata?: nbformat.INotebookMetad
     }
     // If kernel spec is defined & we have a language in that, then use that information.
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    const kernelSpec: IJupyterKernelSpec | undefined = metadata.kernelspec as any;
+    const kernelSpec: Partial<IJupyterKernelSpec> | undefined = metadata.kernelspec as any;
```
Comment (DonJayamanne) [inline src/kernels/helpers.ts:1305]: This was a bug fix, actually the kernelspec is only a partial type here

File: src/kernels/helpers.ts [hunk 2]
```diff
@@ -1302,10 +1302,10 @@ export function getLanguageInNotebookMetadata(metadata?: nbformat.INotebookMetad
...
     return getLanguageInKernelSpec(kernelSpec) || metadata.language_info?.name;
 }
-export function getLanguageInKernelSpec(kernelSpec?: IJupyterKernelSpec | undefined): string | undefined {
+export function getLanguageInKernelSpec(kernelSpec?: Partial<IJupyterKernelSpec> | undefined): string | undefined {
```
Comment (DonJayamanne) [inline src/kernels/helpers.ts:1308]: This was a bug fix, actually the kernelspec is only a partial type here

File: src/notebooks/helpers.ts [hunk 3]
```diff
@@ -102,12 +102,12 @@ export async function updateNotebookDocumentMetadata(
 
 export function isPythonNotebook(metadata?: nbformat.INotebookMetadata) {
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    const kernelSpec = metadata?.kernelspec as any as IJupyterKernelSpec | undefined;
+    const kernelSpec = metadata?.kernelspec as any as Partial<IJupyterKernelSpec> | undefined;
```
Comment (DonJayamanne) [inline src/notebooks/helpers.ts:105]: This was a bug fix, actually the kernelspec is only a partial type here

File: src/notebooks/helpers.ts [hunk 4]
```diff
@@ -102,12 +102,12 @@ export async function updateNotebookDocumentMetadata(
...
     if (metadata?.language_info?.name && metadata.language_info.name !== PYTHON_LANGUAGE) {
         return false;
     }
 
-    if (kernelSpec?.name.includes(PYTHON_LANGUAGE)) {
+    if (kernelSpec?.name?.includes(PYTHON_LANGUAGE)) {
```
Comment (DonJayamanne) [inline src/notebooks/helpers.ts:110]: This was failing, as `name` could be nullable.

File: src/webviews/webview-side/ipywidgets/kernel/index.ts [hunk 5]
```diff
@@ -55,7 +55,7 @@ class WidgetManagerComponent {
             cdnsUsed: this.widgetsCanLoadFromCDN,
             isOnline: data.isOnline,
             timedout: data.timedout,
-            error: data.error
+            error: JSON.stringify(data.error)
```
Comment (DonJayamanne) [inline src/webviews/webview-side/ipywidgets/kernel/index.ts:58]: Changes to ensure all of the error information is sent, if we don't do this at the extension host end we get `[Object object]` & don't get any of the error information in the logs.

**Review by rchiodo — APPROVED** [review]

**Pull request merged**